Geometry type schema is now more flexible

snorfalorpagus opened the pull request:
This PR modifies the validation done when writing geometries to ensure they match the collection schema.

The existing behaviour of specifying a single geometry type is unchanged:

```
fiona.open(..., schema={"geometry": "Point", ...})
```

Alternatively you can specify a list of types:

```
fiona.open(..., schema={"geometry": ("Point", "LineString", "Polygon", ...})
```

Or allow any type:

```
fiona.open(..., schema={"geometry": "Unknown", ...})
```

This works for the edge cases I could think of (e.g. ESRI Shapefile doesn't care between LineString/MultiLineString, Polygon/MultiPolygon), but I may have missed something out.

A new exception is raises if a geometry is written that doesn't match the schema: `GeometryTypeValidationError`.

It's possible to specify a schema that isn't actually valid for the driver. For example, a schema of `("Point", "LineString")` for an ESRI Shapefile. The code doesn't check for this explicitly, but will raise a `RuntimeError` if you actually try to write something invalid.

Closes #446.

sgillies commented on fiona/collection.py:
```diff
@@ -427,6 +430,30 @@ def __del__(self):
+        valid_types.remove("Unknown")
+
+    if driver == "ESRI Shapefile" and "Point" not in valid_types:
+        print("OMG")
```
Funny, but I think we should remove this :)

snorfalorpagus replied:
Ha! This wasn't supposed to be left in. It's my go-to statement for checking branches are being run. I'll remove it.

sgillies commented on fiona/collection.py:
```diff
@@ -427,6 +430,30 @@ def __del__(self):
+    for geom_type in schema_geom_type:
+        valid_types.add(str(geom_type).lstrip("3D "))
+
+    if "Unknown" in valid_types:
```
I think "Any" might be better than "Unknown".

snorfalorpagus replied:
The reason for using "Unknown" is that this is what GDAL uses when it opens a multi-type layer. I think we need to support it, so that you can reuse the schema when reading from one layer directly into another - I'll add a test specifically for this. I don't mind adding an alias for this for writing layers. "Any" is my preference.

sgillies replied:
Ah, okay, I see it does. I quibble with GDAL's semantics all the time. Consider the two statements:

1. Create a new layer to store features with an *unknown* geometry type
2. Create a new later to store features with *any* geometry type

I think *any* suits the writing situation better and *unknown* suits the reading situation better, if you see what I mean. I think we should choose one and follow GDAL here. "unknown" it is.

sgillies commented on fiona/errors.py:
```diff
@@ -33,5 +33,9 @@ class UnsupportedGeometryTypeError(KeyError):
 
 
+class GeometryTypeValidationError(FionaValueError):
+    """Tried to write a geometry type not specified in the schema"""
```
Great.

snorfalorpagus commented:
I've added "Any" as an alias for "Unknown" when creating layers. "Unknown" will still work, and needs to so that we can re-use the metadata from a source layer in the destination layer (see `test_unknown`).

I fixed a bug in `geometry_type_code` whereby the exception was being ignored by cython.

I also added a check to see if the session `isactive` before trying to close it (in `__del__`) which was triggering a warning when a layer failed to open due to an invalid geometry type in the schema.

You'll need to squash the commits when you do the merge to avoid the `print("OMG")` ending up in the history.

snorfalorpagus commented:
This may need some more work in relation to #465.

Currently if you open a layer with schema="3D Point" and try to write a geometry type="Point", it works but uses a 0 value for the Z coordinate. I'm not sure what the correct behaviour should be.

sgillies commented:
@snorfalorpagus the options would be to use 0 for z or refuse to convert and raise an exception, yes?

Elsewhere, I've written that users shouldn't depend on Fiona schemas as a crosswalk for their data. That you shouldnt' specify `float:7.2` and expect a string like `'123.45678'` to be converted perfectly.

That said, adding a z value to 2D points is the user-friendly thing to do, and `0` is the best of all possible values.

snorfalorpagus commented:
I'm OK with setting the z-value to 0 instead of raising an exception so long as it's documented.

I'm looking into the error for `test_schema_geom.py::test_unknown` now - weirdly it passes on my machine.

snorfalorpagus commented:
I can reproduce the issue now. It looks like the branch is OK and the issue is in the merge to master.

snorfalorpagus commented:
I've tracked the issue down to this:

https://github.com/Toblerity/Fiona/blame/master/fiona/ogrext.pyx#L257-L262

Previously if a feature had no geometry set the fiona feature would contain `{"geometry": None}`. The new behaviour (#488) removes the key completely if the geometry is missing. This works until you try to write the feature to another layer. I see two options here: either revert the change to handling of null geometry (i.e. use `"geometry": None`), or interpret the missing key as a null geometry when writing features. @sgillies?

sgillies commented:
@snorfalorpagus after line 262, let's add a `else: fiona_feature["geometry"] = None` to get back to the original behavior. Okay?

snorfalorpagus commented:
The failures seem to be GitHub related...

snorfalorpagus commented:
I think I confused Travis by accidentally pushing a branch with the same name to origin, rather than my own user. The PR tests are passing which is what matters. I think this is ready now.
